src/io/sendmsg.rs

```diff
@@ -0,0 +1,118 @@
+    fd: SharedFd,
+    #[allow(dead_code)]
+    io_bufs: Vec<T>,
+    #[allow(dead_code)]
+    io_slices: Vec<IoSlice<'static>>,
+    #[allow(dead_code)]
+    socket_addr: Option<Box<SockAddr>>,
+    msg_control: Option<U>,
+    msghdr: libc::msghdr,
+
+    /// Hold the number of transmitted bytes
+    bytes: usize,
+}
+
+impl<T: BoundedBuf, U: BoundedBuf> Op<SendMsg<T, U>, MultiCQEFuture> {
+    pub(crate) fn sendmsg(
+        fd: &SharedFd,
+        io_bufs: Vec<T>,
+        socket_addr: Option<SocketAddr>,
+        msg_control: Option<U>,
+    ) -> io::Result<Self> {
+        use io_uring::{opcode, types};
+
+        let mut msghdr: libc::msghdr = unsafe { std::mem::zeroed() };
+
+        let mut io_slices: Vec<IoSlice<'static>> = Vec::with_capacity(io_bufs.len());
+
+        for io_buf in &io_bufs {
+            io_slices.push(IoSlice::new(unsafe {
+                std::slice::from_raw_parts(io_buf.stable_ptr(), io_buf.bytes_init())
+            }))
+        }
+
+        msghdr.msg_iov = io_slices.as_ptr() as *mut _;
+        msghdr.msg_iovlen = io_slices.len() as _;
+
+        let socket_addr = match socket_addr {
+            Some(_socket_addr) => {
+                let socket_addr = Box::new(SockAddr::from(_socket_addr));
+                msghdr.msg_name = socket_addr.as_ptr() as *mut libc::c_void;
+                msghdr.msg_namelen = socket_addr.len();
+                Some(socket_addr)
+            }
+            None => {
+                msghdr.msg_name = std::ptr::null_mut();
+                msghdr.msg_namelen = 0;
+                None
+            }
+        };
+
+        match msg_control {
+            Some(ref _msg_control) => {
+                msghdr.msg_control = _msg_control.stable_ptr() as *mut _;
+                msghdr.msg_controllen = _msg_control.bytes_init();
+            }
+            None => {
+                msghdr.msg_control = std::ptr::null_mut();
+                msghdr.msg_controllen = 0_usize;
+            }
+        }
+
+        CONTEXT.with(|x| {
+            x.handle().expect("Not in a runtime context").submit_op(
+                SendMsg {
+                    fd: fd.clone(),
+                    io_bufs,
+                    socket_addr,
+                    io_slices,
+                    msg_control,
+                    msghdr,
+                    bytes: 0,
+                },
+                |sendmsg| {
+                    opcode::SendMsg::new(
+                        types::Fd(sendmsg.fd.raw_fd()),
+                        &sendmsg.msghdr as *const _,
+                    )
+                    .build()
+                },
+            )
+        })
+    }
+}
+
+impl<T, U> Completable for SendMsg<T, U> {
+    type Output = (io::Result<usize>, Vec<T>, Option<U>);
+
+    fn complete(self, cqe: CqeResult) -> (io::Result<usize>, Vec<T>, Option<U>) {
+        // Convert the operation result to `usize`, and add previous byte count
+        let res = cqe.result.map(|v| self.bytes + v as usize);
```

Comment:
The cqe.result will be the number of bytes written, so no need to use map and expect to add anything. But you wouldn't have figured that out from my other comments. I just wanted to help avoid any possible confusion.

Reply:
Thanks!

Reply:
I'm just curious: why does CqeResult give an u32? Does that mean io-uring doesn't handle slices longer than 4GB?

Reply:
Yes, almost. It actually comes from a signed 32 bit field, so 2GB will be the largest value that can be returned. I don't expect you can get that changed but if you want to ask for background, the team at liburing seems to be open to answering questions about io_uring. Here's is the relevant field in their structure.

https://github.com/axboe/liburing/blob/9134244a4e16e68cb514ea5192fcd7e9beca96f5/src/include/liburing/io_uring.h#L358

Reply:
Thanks!